File: backend/app/models/delivery_request.py
import uuid
from datetime import datetime

from app.extensions import db
from app.utils.datetime_format import utc_isoformat


class DeliveryRequest(db.Model):
    __tablename__ = "delivery_requests"

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    customer_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("users.id"), nullable=False, index=True)
    delivery_address_id = db.Column(
        db.UUID(as_uuid=True), db.ForeignKey("delivery_addresses.id"), nullable=False
    )
    status = db.Column(db.String(20), nullable=False, default="pending", index=True)
    delivery_fee_jmd = db.Column(db.Numeric(12, 2), nullable=False)
    notes = db.Column(db.String(500))
    requested_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    completed_at = db.Column(db.DateTime)
    completed_by_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("users.id"))
    in_progress_at = db.Column(db.DateTime)
    in_progress_by_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("users.id"))
    cancelled_at = db.Column(db.DateTime)

    customer = db.relationship("User", foreign_keys=[customer_id], backref="delivery_requests")
    delivery_address = db.relationship("DeliveryAddress", backref="delivery_requests")
    completed_by = db.relationship("User", foreign_keys=[completed_by_id])
    in_progress_by = db.relationship("User", foreign_keys=[in_progress_by_id])
    package_links = db.relationship(
        "DeliveryRequestPackage",
        backref="delivery_request",
        lazy=True,
        cascade="all, delete-orphan",
    )

    def to_dict(self, include_packages: bool = False, include_address: bool = True) -> dict:
        from app.constants import DELIVERY_REQUEST_STATUS_LABELS

        data = {
            "id": str(self.id),
            "customer_id": str(self.customer_id),
            "delivery_address_id": str(self.delivery_address_id),
            "status": self.status,
            "status_label": DELIVERY_REQUEST_STATUS_LABELS.get(self.status, self.status),
            "delivery_fee_jmd": float(self.delivery_fee_jmd),
            "notes": self.notes,
            "requested_at": utc_isoformat(self.requested_at),
            "completed_at": utc_isoformat(self.completed_at),
            "completed_by_name": self.completed_by.full_name if self.completed_by else None,
            "in_progress_at": utc_isoformat(self.in_progress_at),
            "in_progress_by_name": self.in_progress_by.full_name if self.in_progress_by else None,
            "cancelled_at": utc_isoformat(self.cancelled_at),
            "package_count": len(self.package_links),
        }
        if include_address and self.delivery_address:
            data["delivery_address"] = self.delivery_address.to_dict()
        if include_packages:
            data["packages"] = [link.to_dict(include_package=True) for link in self.package_links]
        if self.customer:
            data["customer_name"] = self.customer.full_name
            data["shipping_id"] = self.customer.shipping_id
        return data


class DeliveryRequestPackage(db.Model):
    __tablename__ = "delivery_request_packages"

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    delivery_request_id = db.Column(
        db.UUID(as_uuid=True),
        db.ForeignKey("delivery_requests.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    package_id = db.Column(db.UUID(as_uuid=True), db.ForeignKey("packages.id"), nullable=False, unique=True)

    package = db.relationship("Package", backref="delivery_request_links")

    def to_dict(self, include_package: bool = False) -> dict:
        data = {
            "id": str(self.id),
            "delivery_request_id": str(self.delivery_request_id),
            "package_id": str(self.package_id),
        }
        if include_package and self.package:
            from app.constants import STATUS_LABELS

            data["tracking_number"] = self.package.tracking_number
            data["total_due_jmd"] = (
                float(self.package.total_due_jmd) if self.package.total_due_jmd is not None else None
            )
            data["billing_status"] = self.package.billing_status
            data["status"] = self.package.status
            data["status_label"] = STATUS_LABELS.get(self.package.status, self.package.status)
        return data

File: server/src/lib/serializers/deliveryRequest.ts
import { DELIVERY_REQUEST_STATUS_LABELS, STATUS_LABELS } from '../../constants.js'
import type {
  deliveryAddresses,
  deliveryRequestPackages,
  deliveryRequests,
  packages,
  users,
} from '../../db/schema/index.js'
import { utcIsoformat } from '../dates.js'
import { deliveryAddressToDict } from './deliveryAddress.js'
import { userFullName } from './user.js'

type DeliveryRequestRow = typeof deliveryRequests.$inferSelect
type DeliveryRequestPackageRow = typeof deliveryRequestPackages.$inferSelect
type PackageRow = typeof packages.$inferSelect
type DeliveryAddressRow = typeof deliveryAddresses.$inferSelect
type UserRow = typeof users.$inferSelect

export function deliveryRequestPackageToDict(
  link: DeliveryRequestPackageRow,
  pkg?: PackageRow | null,
): Record<string, unknown> {
  const data: Record<string, unknown> = {
    id: link.id,
    delivery_request_id: link.deliveryRequestId,
    package_id: link.packageId,
  }
  if (pkg) {
    data.tracking_number = pkg.trackingNumber
    data.total_due_jmd = pkg.totalDueJmd != null ? parseFloat(pkg.totalDueJmd) : null
    data.billing_status = pkg.billingStatus
    data.status = pkg.status
    data.status_label = STATUS_LABELS[pkg.status] ?? pkg.status
  }
  return data
}

export function deliveryRequestToDict(
  request: DeliveryRequestRow,
  opts: {
    includePackages?: boolean
    includeAddress?: boolean
    packageLinks?: Array<{ link: DeliveryRequestPackageRow; pkg?: PackageRow | null }>
    deliveryAddress?: DeliveryAddressRow | null
    customer?: UserRow | null
    completedBy?: UserRow | null
    inProgressBy?: UserRow | null
  } = {},
): Record<string, unknown> {
  const includeAddress = opts.includeAddress !== false
  const data: Record<string, unknown> = {
    id: request.id,
    customer_id: request.customerId,
    delivery_address_id: request.deliveryAddressId,
    status: request.status,
    status_label: DELIVERY_REQUEST_STATUS_LABELS[request.status] ?? request.status,
    delivery_fee_jmd: parseFloat(request.deliveryFeeJmd),
    notes: request.notes,
    requested_at: utcIsoformat(request.requestedAt),
    completed_at: utcIsoformat(request.completedAt),
    completed_by_name: opts.completedBy ? userFullName(opts.completedBy) : null,
    in_progress_at: utcIsoformat(request.inProgressAt),
    in_progress_by_name: opts.inProgressBy ? userFullName(opts.inProgressBy) : null,
    cancelled_at: utcIsoformat(request.cancelledAt),
    package_count: opts.packageLinks?.length ?? 0,
  }
  if (includeAddress && opts.deliveryAddress) {
    data.delivery_address = deliveryAddressToDict(opts.deliveryAddress)
  }
  if (opts.includePackages && opts.packageLinks) {
    data.packages = opts.packageLinks.map(({ link, pkg }) => deliveryRequestPackageToDict(link, pkg))
  }
  if (opts.customer) {
    data.customer_name = userFullName(opts.customer)
    data.shipping_id = opts.customer.shippingId
  }
  return data
}
